Add per-concept noise and timesteps distribution settings

FennelFetish opened the pull request:
This PR adds support for per-concept noise and timestep distribution settings:

**UI Changes:**
- Refactor `TimestepDistributionWindow` into a separate frame so the controls and the plot can be reused in the `ConceptWindow`.
  - I also moved the "*Update Preview*" button to the left, because it was always hidden when opening the window.
- Added "*noise*" tab to ConceptWindow.
  - It includes separate toggles for "*Noise Override*" and "*Timestep Distribution Override*".
  - The concept settings are not tied to a specific model type, so it displays all settings and doesn't check for compatibility with the model.
  - The timestep distribution plot is left empty when opening the `ConceptWindow` because it would delay the loading too much.

**Config Changes:**
- Added `ConceptNoiseConfig` with the same names for the settings as in `TrainConfig`.

**Pipeline Changes:**
The goal was to get per-sample concept information for a batch. Passing whole dicts to the batching function seemed inefficient and object references don't work. Instead, I added `concept_index` (int) to the output names so the `ConceptConfig` can be retrieved from `TrainConfig` inside the training loop.
Those indexes are assembled by a modified `CollectPaths` [in my MGDS fork here](https://github.com/Nerogar/mgds/compare/master...FennelFetish:mgds:concept-indexes).

The `TrainConfig` was available inside the training loop, but `config.concepts` was None.
So I modified `CollectPaths` to also store the enabled concepts (dicts) and provide them as meta information.
I added a pipeline module (`StoreConcepts`) which retrieves these meta-concepts and stores them in `config.concepts`.

Then I changed `DataLoaderMgdsMixin` so it always reloads the concepts at the beginning of the pipeline, without checking if they already exist. This is needed so `StoreConcepts` can update the settings when concepts are modified between training runs.

Getting the concept of the samples inside the training loop:
`batch_concepts = [config.concepts[i] for i in batch['concept_index']]`

But I have the impression that the way the concepts are set to `TrainConfig` isn't robust and could cause issues. In my testing, `config.concepts` was always None, yet `DataLoaderMgdsMixin` checked for its presence.
Are the concepts conditionally set by something else?

**Noise Changes:**
- Use the concept settings in `ModelSetupNoiseMixin` for noise shaping and timestep distribution sampling.
- The `__get_concept_noise_configs` helper function creates a list with configs.
  - If override is enabled, it uses `ConceptNoiseConfig`. Otherwise it defaults to `TrainConfig`.
- If any sample of a batch has overriding enabled, it calls `_create_noise()` or `_get_timestep_...()` separately, once for each sample.
  - Could be optimized by grouping the samples by config and calling the functions only once per group.
  - The noise for a whole batch is still generated in one call, so I modified `_create_noise` to use in-place addition.

I left some debug printing in `BaseStableDiffusionXLSetup`, in case anyone wants to check.

The per-concept noise settings are only used for SDXL training so far.
Please let me know if this is the right approach and then I can add it to the other models too.

References:
[[Feat]: move the noising arguments (offset, perturbation, min, max, weight, bias) to concepts](https://github.com/Nerogar/OneTrainer/issues/684)

O-J1 commented:
Not yet tested, 2 quick questions

1. How does this handle users who want just normal dataset wide noise? Just an override?
2. This is a question for both of you, what are your thoughts on addressing this comment by MX regarding per image configuration https://github.com/Nerogar/OneTrainer/issues/238#issuecomment-2041001425 (former dev, legal issues stopped him)

FennelFetish commented:
1. Two separate overrides for noise and timestep distribution. Yes, if it's off (default), it takes the global, dataset wide settings.

2. I'd love an option to set the repeats per image. I think it's the implementation of the specific overrides that would need quite some work, but that could be done step by step.

What I'm doing right now for the per-image repeats:
- Tag the images with the repeat number, grouped into 0.3, 0.5, 1.0
- Filter and rearrange the files into new folders with symlinks.
- Load these folders as concepts in OT and set the repeats per-concept.
(I'm content with this solution though)

I think per-concept overrides would still be useful because setting the values per image also takes time when building datasets.

The data loading pipeline could allow for both:
image settings > concept settings > global settings

Also, some settings could shift some distributions in unforeseen ways. Like the timestep distribution. Plots for viewing global distributions with all the per-concept and per-file overrides included would be nice (it would need to load all files).

For passing the overridden settings per sample to the training loop, it could possibly work with a custom collate function in MGDS' `TrainDataLoader`: https://docs.pytorch.org/docs/stable/data.html#torch.utils.data.default_collate

```python
def custom_collate(sample_data: list[dict]):
    # keep the dicts with the overrides unchanged
    config_overrides: list[dict] = [data.pop("config_overrides") for data in sample_data]
    batch = default_collate(sample_data)
    batch["config_overrides"] = config_overrides  
    return batch
```
That's probably nonsense code, but if the idea works, that could make it easier to pass overrides into the training loop. Instead of what I've suggested with the `StoreConcepts` pipeline module and `concept_index`. I was concerned about performance of the collate_fn which is executed on each training step and would need to handle all additional keys.

Goldenkoron commented:
> Thanks for this PR again! I had the chance to take a first look at the code now, but first regarding your discussion above:
> 
> 1. What would be the purpose of per-image settings for timestep distribution and noise settings?


Not the dev for this, but my personal use case would be to separate datasets by desired composition and desired style/detail.

For instance a very diverse dataset focusing primarily the high timesteps, and a higher aesthetic but significantly less diverse dataset for low timesteps to improve the style and details without overfitting the composition.

FennelFetish commented:
Thanks for the review.

1. I agree that the redeclaration was ugly. But changing the parameter name like that would expose details about the function that should be irrelevant to a caller. Instead, I changed the internal name to `batch_shift`.
I hope that's okay.

2. Done

3. The tolist() is indeed the slowest part of it (still <100 µs total with default settings), but the keys shouldn't cause issues with memory. Each key consists of 5 floats, created from tensors on the CPU device. Number of keys = batch size.
The cache typically has one entry with the default settings, plus one for each concept with overrides. I'd expect that only with per-file overrides or with dynamic timestep shifting and many different resolutions the cache would grow larger. And then it takes 4KB+ of VRAM per entry.
Pre-hashing the key-tuples could lead to collisions. But it could make sense to round the values to a few digits to share cache entries when the settings are very close.

4. I expanded the comment to better explain where the maximum scale value comes from. And added a plot.

---

  1. `prepare_batch_config()`: That's right. `None` values come from empty input fields in the UI.

dxqb commented on modules/dataLoader/mixin/DataLoaderText2ImageMixin.py:
```diff
@@ -303,6 +308,21 @@ def _output_modules_from_out_names(
         return modules
+
+    @staticmethod
+    def __create_prepare_batch_config(config: TrainConfig):
```
I dislike the grouping because it's not supported by the dataloader collate_fn but you have to group yourself then, which makes your code more complicated.

dxqb commented:
when you finish this PR, could you please consider merging in https://github.com/Nerogar/OneTrainer/pull/1225?
If I merge it now I will force a conflict. It seems to me it's better to coordinate this one in advance
